presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/FailureAwareHiveMetaStoreClient.java

```diff
@@ -0,0 +1,358 @@
+    public List<Partition> getPartitionsByNames(String databaseName, String tableName, List<String> partitionNames)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.getPartitionsByNames(databaseName, tableName, partitionNames));
+    }
+
+    @Override
+    public List<Role> listRoles(String principalName, PrincipalType principalType)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.listRoles(principalName, principalType));
+    }
+
+    @Override
+    public List<HiveObjectPrivilege> listPrivileges(String principalName, PrincipalType principalType, HiveObjectRef hiveObjectRef)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.listPrivileges(principalName, principalType, hiveObjectRef));
+    }
+
+    @Override
+    public List<String> getRoleNames()
+            throws TException
+    {
+        return runWithHandle(delegate::getRoleNames);
+    }
+
+    @Override
+    public void createRole(String role, String grantor)
+            throws TException
+    {
+        runWithHandle(() -> delegate.createRole(role, grantor));
+    }
+
+    @Override
+    public void dropRole(String role)
+            throws TException
+    {
+        runWithHandle(() -> delegate.dropRole(role));
+    }
+
+    @Override
+    public boolean grantPrivileges(PrivilegeBag privilegeBag)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.grantPrivileges(privilegeBag));
+    }
+
+    @Override
+    public boolean revokePrivileges(PrivilegeBag privilegeBag)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.revokePrivileges(privilegeBag));
+    }
+
+    @Override
+    public void grantRole(String role, String granteeName, PrincipalType granteeType, String grantorName, PrincipalType grantorType, boolean grantOption)
+            throws TException
+    {
+        runWithHandle(() -> delegate.grantRole(role, granteeName, granteeType, grantorName, grantorType, grantOption));
+    }
+
+    @Override
+    public void revokeRole(String role, String granteeName, PrincipalType granteeType, boolean grantOption)
+            throws TException
+    {
+        runWithHandle(() -> delegate.revokeRole(role, granteeName, granteeType, grantOption));
+    }
+
+    @Override
+    public List<RolePrincipalGrant> listRoleGrants(String name, PrincipalType principalType)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.listRoleGrants(name, principalType));
+    }
+
+    @Override
+    public void setUGI(String userName)
+            throws TException
+    {
+        runWithHandle(() -> delegate.setUGI(userName));
+    }
+
+    public interface ResponseHandle
+    {
+        void success();
+
+        void failed(TException e);
+    }
+
+    private <T> T runWithHandle(ThrowingSupplier<T> supplier)
+            throws TException
+    {
+        T t;
+        try {
+            t = supplier.get();
```

Comment:
Make `responseHandle.success();` part of `try {` as it cannot throw `TException`, e.g:
```
try {
  T result = supplier.get();
  responseHandle.success();
  return result;
} catch(TException e) {
  ..
}
```
```